no/#1/in-memory-crud-user

OtiliaN opened the pull request:
- CRUD for User
- Tests for Repo, Service and Controller

OtiliaN committed:
#1 - finished first feature

OtiliaN committed:
#1 - added annotation for api error responses

Andrei0921 commented on backend/src/main/java/cloudflight/integra/backend/dto/UserDto.java:
```diff
@@ -0,0 +1,11 @@
+import java.time.LocalDateTime;
+
+public class UserDto {
+    public Long id;
```
i think it would be better if you make the fields private and use getters and setters to maintain encapsulation

Andrei0921 commented on backend/src/main/java/cloudflight/integra/backend/repository/UserRepositoryInMemoryImpl.java:
```diff
@@ -0,0 +1,45 @@
+
+@Repository
+public class UserRepositoryInMemoryImpl implements IUserRepository<Long, User> {
+    private final Map<Long, User> users = new HashMap<>();
```
you can use AtomicLong and ConcurrentHashMap so the repository it would be thread safe

Andrei0921 commented on backend/src/main/java/cloudflight/integra/backend/repository/UserRepositoryInMemoryImpl.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public User findOne(Long id) {
```
you can use Optional here to avoid null pointer exception

Andrei0921 commented on backend/src/main/java/cloudflight/integra/backend/controller/problem/UserRestExceptionHandler.java:
```diff
@@ -0,0 +1,39 @@
+     * If the exception message contains "not found", returns a 404 Not Found response.
+     * Otherwise, returns a 500 Internal Server Error response with the exception message.
+     * */
+    @ExceptionHandler(RuntimeException.class)
```
you can make a custom NotFoundException to make the code cleaner

the-coding-cloud commented on backend/src/main/java/cloudflight/integra/backend/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/users")
```
as discussed earlier, let's try to use the /api/v1/users convention to distinguish between frontend paths and api paths

the-coding-cloud commented on backend/src/main/java/cloudflight/integra/backend/entity/validator/UserValidator.java:
```diff
@@ -0,0 +1,54 @@
+        }
+
+        if (!errors.isEmpty()) {
+            throw new ValidationException(String.join(" ", errors));
```
have a list of errors as a field inside the validation exception instead of using just one big message

OtiliaN committed:
#1 - refactor: improve encapsulation, thread-safety and error handling
- made fields private and added getters/setters for encapsulation in UserDto
- replaced data structures with AtomicLong and ConcurrentHashMap for thread safety
- used Optional in Repo to avoid NullPointerExceptions
- added custom NotFoundException
- updated API paths to follow /api/v1/users convention
- changed ValidationException to store a list of errors instead of a single message

the-coding-cloud reviewed (approved):
Good job!
